gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -442,4 +442,23 @@ private void saveDeviceAndRedirect(RoutingContext routingContext, Client client,
             ).subscribe();
         }
     }
+
+    private boolean enableAlternateMFAOptions(Client client, io.gravitee.am.model.User endUser ){
+        if(endUser.getFactors() == null || endUser.getFactors().size() <= 1) {
+            return false;
+        }
+        else if(client.getFactors() == null || client.getFactors().size() <= 1){
+            return false;
+        }
+        else {
+            final Set<String> clientFactorIds = client.getFactors();
+            final List<EnrolledFactor> activeEnrolledFactors = endUser.getFactors()
+                    .stream()
+                    .filter(enrolledFactor -> factorManager.get(enrolledFactor.getFactorId()) != null)
+                    .filter(enrolledFactor -> clientFactorIds.contains(enrolledFactor.getFactorId()))
+                    .collect(Collectors.toList());
```

Comment:
Maybe the `getEnabledFactors` could serve its purpose in a utility class ?

Reply:
discussed with @remisultan and decided to leave it as it is for now.